scipy/integrate/_tanhsinh.py

```diff
@@ -0,0 +1,581 @@
+        arXiv preprint arXiv:2007.15057 (2020).
+    [3] van Engelen, Robert A.  "Improving the Double Exponential Quadrature
+        Tanh-Sinh, Sinh-Sinh and Exp-Sinh Formulas."
+        https://www.genivia.com/files/qthsh.pdf
+    [4] Takahasi, Hidetosi, and Masatake Mori. "Double exponential formulas for
+        numerical integration." Publications of the Research Institute for
+        Mathematical Sciences 9.3 (1974): 721-741.
+
+    Example
+    -------
+    Evaluate the Gaussian integral:
+
+    >>> import numpy as np
+    >>> from scipy import integrate
+    >>> def f(x):
+    ...     return np.exp(-x**2)
+    >>> res = _tanhsinh(f, -np.inf, np.inf)
+    >>> res.integral  # true value is np.sqrt(np.pi), 1.7724538509055159
+    1.772453850905516
+    >>> res.error  # actual error is ~2.220446049250313e-16
+    1.062283726062058e-15
+
+    The value of the Gaussian function (bell curve) is nearly zero for
+    arguments sufficiently far from zero, so the value of the integral
+    over a finite interval is nearly the same.
+
+    >>> _tanhsinh(f, -20, 20).integral
+    1.7724538509055152
+
+    However, with unfavorable integration limits, the integration scheme
+    may not be able to find the important region.
+
+    >>> _tanhsinh(f, -np.inf, 20).integral
+    1.967881377548732e-19
+
+    In such cases, or when there are singularities within the interval,
+    break the integral into parts with endpoints at the important points.
+
+    >>> _tanhsinh(f, -np.inf, 0).integral + _tanhsinh(f, 0, 20).integral
+    1.7724538509055163
+
+    """
+    # Input validation and standardization
+    res = _tanhsinh_iv(f, a, b, log, maxfun, maxlevel,
+                       minlevel, atol, rtol)
+    (f, a, b, log, maxfun, maxlevel, minlevel,
+     atol, rtol, feval_factor) = res
+
+    # Initialization
+    Sk = []  # sequence of integral estimates for error estimation
+    # Most extreme abscissae and corresponding `fj`s, `wj`s in Euler-Maclaurin
+    # sum. These are dummy values that will be replaced in the first iteration.
+    last_terms = (np.inf, np.nan, 0, -np.inf, np.nan, 0, 0)
+    Sn = aerr = np.nan  # integral and error are NaN until determined otherwise
+    status = -1  # "Iteration in progress." for callback
+    feval = 0  # function evaluation counter
+    h0 = _get_step()
+
+    for n in range(minlevel, maxlevel+1):  # for each "level"
+        h = h0 / 2**n  # step size
+
+        # retrieve abscissa-weight pairs from cache
+        xjc, wj = _get_pairs(n, h0, inclusive=(n==minlevel))
+
+        # Determine whether evaluating the function at the abscissae will
+        # cause the function evaluation limit to be exceeded. If so, break.
+        # Otherwise, increment the function eval counter.
+        if feval + 2*len(xjc) * feval_factor > maxfun:
+            status = 2
+            break
+
+        # Transform the abscissae as required by the limits of integration
+        xj, wj = _transform_to_limits(a, b, xjc, wj)
```

Comment:
```suggestion
        xj, wj = _transform_to_limits(xjc, wj, a, b)
```
Would be more intuitive, I think.